Schema: Initial Azure Monitor query schema

aangelisc opened the pull request:
This PR contains the initial schema definition and types for Azure Monitor.

Some questions/points worth noting at the moment:

- Consistently got the error: `may only unify a referenced enum with a concrete literal member of that enum`. Honestly not 100% sure how I fixed it (although I *think* I understand what it means). I believe this error could do with some better wording for clarification.
- *ISSUE*: Plugins with spaces in names have TS incorrectly generated e.g. `Azure Monitor`.
- Can we have some types where only TS code is generated and not Go? There is a type that exists in the front-end and not the back-end. It is deprecated and I guess we shouldn't really have circumstances where the frontend and backend query types differ from each-other.
- Is it possible to prevent the default interface generation (TS)?
- *ISSUE*: In Go a struct isn’t referenced in a parent struct even if it’s defined. The whole struct is instead duplicated.

Further notes:

This PR can't be merged yet because of naming issues in the backend Azure Monitor plugin. The folder contained within `tsdb` is called `azuremonitor` where it seems the expected name is `grafana-azure-monitor-datasource`. I'm assuming this needs to be updated.

I need to update the relevant code in the front-end and back-end to test the types and validate that they function. 

I think there's excessive boilerplate generated for the backend that isn't required but that's not a big issue atm.

As pretty much all of the properties are optional this leads to the properties of the Go types being pointers. This may cause issues and require more refactoring..

Also still need to schematise the `dataSourceCfg`

aangelisc committed:
Initial schema

- Add types based off of current frontend

Clarity-89 reviewed (commented):
> Can we have some types where only TS code is generated and not Go? There is a type that exists in the front-end and not the back-end. It is deprecated and I guess we shouldn't really have circumstances where the frontend and backend query types differ from each-other.

That's not possible and is deliberate. What can be done though, is to import the type into `types.ts`, add the needed type or property and re-export that type. [Here's an example.](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/geomap/types.ts#L13-L15) Also in such case it's better to [re-export](https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/geomap/types.ts#L64) all types from `models.gen.ts` there, so it becomes the unified source of TS types.

Clarity-89 commented:
> As pretty much all of the properties are optional this leads to the properties of the Go types being pointers. This may cause issues and require more refactoring..

Maybe [this](https://github.com/grafana/grafana/pull/61948/) will help there?

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/composable_dataquery.cue:
```diff
@@ -0,0 +1,188 @@
+
+						// Azure Monitor Metrics sub-query properties
+						#AzureMetricQuery: {
+							resources?: [...#AzureMonitorResource]
```
We should probably define these as mandatory. I _think_ that these are optional in the frontend because it's waiting for user input to fill certain fields. This is not really necessary AFAIK. A working query should have `resources`, `metricNamespace`, `metricName`... The model should probably reflect that (and leave as optional really the fields that can be empty).

aangelisc replied:
Agreed on this, I think it's time to take the opportunity and refactor the types in Azure Monitor as a part of the schematisation 😊

sdboyer replied:
> I think that these are optional in the frontend because it's waiting for user input to fill certain fields. 

This strikes right at the heart of one of the really interesting changes inherent in moving to a schema-centric model - when it comes to defaults and optionality, we have to think about object lifecycle quite differently. i wrote a bit about it [here](https://docs.google.com/document/d/1RRPGeD7DQ7wot0Pwnc4UZolXrZYip0ViqOozl9zbBpE/edit#heading=h.fft1p4jlnh9y), but it's going to be an evolving area.

And yes in this case i agree that "waiting for frontend user input" is not a sufficient reason to mark a field as optional

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/composable_dataquery.cue:
```diff
@@ -0,0 +1,188 @@
+							filter?: string
+						} @cuetsy(kind="interface")
+
+						#GrafanaTemplateVariableQueryType: "AppInsightsMetricNameQuery" | "AppInsightsGroupByQuery" | "SubscriptionsQuery" | "ResourceGroupsQuery" | "ResourceNamesQuery" | "MetricNamespaceQuery" | "MetricNamesQuery" | "WorkspacesQuery" | "UnknownQuery" @cuetsy(kind="type")
+						#BaseGrafanaTemplateVariableQuery: {
+							rawQuery?: string
+							...
+						} @cuetsy(kind="interface")
+						#UnknownQuery: #BaseGrafanaTemplateVariableQuery & {
+										kind: "UnknownQuery"
+						}                            @cuetsy(kind="interface")
+						#AppInsightsMetricNameQuery: #BaseGrafanaTemplateVariableQuery & {
+										kind: "AppInsightsMetricNameQuery"
+						}                         @cuetsy(kind="interface")
+						#AppInsightsGroupByQuery: #BaseGrafanaTemplateVariableQuery & {
+									kind:       "AppInsightsGroupByQuery"
+									metricName: string
+						}                    @cuetsy(kind="interface")
+						#SubscriptionsQuery: #BaseGrafanaTemplateVariableQuery & {
+									kind: "SubscriptionsQuery"
+						}                     @cuetsy(kind="interface")
+						#ResourceGroupsQuery: #BaseGrafanaTemplateVariableQuery & {
+									kind:         "ResourceGroupsQuery"
+									subscription: string
+						}                    @cuetsy(kind="interface")
+						#ResourceNamesQuery: #BaseGrafanaTemplateVariableQuery & {
+									kind:            "ResourceNamesQuery"
+									subscription:    string
+									resourceGroup:   string
+									metricNamespace: string
+						}                      @cuetsy(kind="interface")
+						#MetricNamespaceQuery: #BaseGrafanaTemplateVariableQuery & {
+							kind:             "MetricNamespaceQuery"
+							subscription:     string
+							resourceGroup:    string
+							metricNamespace?: string
+							resourceName?:    string
+						} @cuetsy(kind="interface")
```
I believe all of these should be deprecated? (it's okay to leave them if you are just copying what we have in TS code).

aangelisc replied:
Yep they're deprecated, I'll validate that they're unused and remove them (and any associated logic).

sdboyer replied:
If you're able to just pull these out now, awesome - i'm glad that schematization could be a forcing function to clean up some code :)

But know that as we turn on use of multi-version schemas via thema in the coming months, there will be new strategies available for handling deprecation of query types. So if trying to remove these now causes headaches when thinking about e.g. backwards compatibility, i'd say it's fine to hold off because the set of tools you have available is about to change

andresmgot reviewed (commented):
Nice start :+1:

aangelisc commented:
> FWIW, I would rename the TS folder to azuremonitor to match the naming convention of the other plugins.

I think I need confirmation from @sdboyer on this. It looks like the generated files make use of the name defined in the `plugin.json`. Is this the expected format for plugins going forward?

sdboyer commented:
> I think I need confirmation from @sdboyer on this. It looks like the generated files make use of the name defined in the `plugin.json`. Is this the expected format for plugins going forward?

Definitely, yes. One of the implications of moving towards making core plugins not meaningfully different from external plugins is that containing folder name can no longer be meaningful - only plugin id is.

cc @grafana/plugins-platform

sdboyer commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/models.gen.ts:
```diff
@@ -0,0 +1,227 @@
+
+export type GrafanaTemplateVariableQuery = (AppInsightsMetricNameQuery | AppInsightsGroupByQuery | SubscriptionsQuery | ResourceGroupsQuery | ResourceNamesQuery | MetricNamespaceQuery | MetricDefinitionsQuery | MetricNamesQuery | WorkspacesQuery | UnknownQuery);
+
+export interface Azure Monitor {}
```
@spinillos could you work out a simpler repro for this buggy output and file an issue (or just a PR?)

Clearly it's an issue with our name sanitizers somewhere, i'm _guessing_ probably [here](https://github.com/grafana/grafana/blob/3b3059c9ce94515baef5c352d07651a404102bfc/pkg/plugins/plugindef/plugindef.go#L43). The right place to put a test case would be [in here](https://github.com/grafana/grafana/blob/3b3059c9ce94515baef5c352d07651a404102bfc/pkg/plugins/pfs/pfs_test.go#L16)

spinillos replied:
Its because cfg.RootName contains spaces. I checked this sanitizer and it doesn't looks like that its causing this issue.

I fixed like this: https://github.com/grafana/thema/pull/104 but idk if its the correct place to do it.

sdboyer replied:
Underlying issue here was fixed in #62144 - merging main and rerunning `make gen-cue` should fix it.

spinillos commented:
Sync your PR with main. We merged a fix for the exported name.

aangelisc committed:
Rename and field-level comments

aangelisc committed:
Merge branch 'main' into schema/azure-monitor

# Conflicts:
#	pkg/kindsys/report.json

aangelisc committed:
Update report and regenerate files

aangelisc committed:
Rename frontend Azure folder

- Doing this for consistency and to ensure code-generation works
- Update betterer results due to file renames

aangelisc committed:
Remove default and add back enum vals that I deleted

aangelisc committed:
Set workspace prop as optional

aangelisc committed:
Replace template variable types

aangelisc committed:
Connect frontend query types

- Keep properties optional for now to avoid major changes
- Rename AzureMetricResource
- Correctly use ResultFormat

aangelisc committed:
Add TSVeneer decorator

sdboyer commented:
> ISSUE: In Go a struct isn’t referenced in a parent struct even if it’s defined. The whole struct is instead duplicated.

That is, if the schema says:

```cue
Parent: {
    ParentField: string
}
Ours: {
    OurField: string
}
```

then you might expect

```go
type Parent struct {
    ParentField string
}

type Our struct {
    OurField string
}
```

but are instead getting

```go
type Our struct {
    ParentField string
    OurField string
}
```

right?

This is a known issue, though i didn't get a chance to write it up. We _have_ to do...probably...for cross-package references, e.g. for `DataQuery` itself. But because of an outstanding issue with codegen, the only way to make that happen is to inline ALL struct refs, even local ones. That clearly sucks.

aangelisc commented on pkg/tsdb/azuremonitor/kinds/dataquery/types_dataquery_gen.go:
```diff
@@ -0,0 +1,447 @@
+	DimensionFilter *string `json:"dimensionFilter,omitempty"`
+
+	// Filters to reduce the set of data returned. Dimensions that can be filtered on are defined by the metric.
+	DimensionFilters *[]struct {
```
@sdboyer this is what I meant (sorry if my explanation was a bit confusing). This `DimensionFilters` struct is defined above on line 155 as `AzureMetricDimension` so I would expect it to be referenced here but instead the whole `AzureMetricDimension` struct is duplicated.

sdboyer replied:
Sorry, yes, this is the flipside of the thing i was describing, but i put it poorly.

These should clearly just be a reference to the other generated type. But they're getting _incorrectly inlined_ right now because of the config option we have to set in order to _correctly inline_ `DataQuery` itself.

It's a bug, and we'll fix it 😄 

(also ughhh we've got slice pointers again! @spinillos )

aangelisc committed:
Merge branch 'main' into schema/azure-monitor

# Conflicts:
#	pkg/kindsys/report.json
#	public/app/plugins/datasource/azuremonitor/components/ResourcePicker/Advanced.test.tsx
#	public/app/plugins/datasource/azuremonitor/components/ResourcePicker/Advanced.tsx
#	public/app/plugins/datasource/azuremonitor/components/ResourcePicker/ResourcePicker.tsx
#	public/app/plugins/datasource/grafana-azure-monitor-datasource/types/query.ts

aangelisc committed:
Update schema

aangelisc committed:
Merge branch 'main' into schema/azure-monitor

# Conflicts:
#	pkg/kindsys/report.json

aangelisc committed:
Update type

aangelisc commented:
Marking this as ready for review with the following caveats:

- Only the frontend types have been hooked up. Backend types aren't yet pending a refactor of types to ensure required properties are no longer optional and that the slice pointers issue is fixed @sdboyer.
- In another PR there will be a larger refactor of the Azure Monitor types to remove any deprecated types, update required properties, and schematise the datasource configuration.

aangelisc committed:
Update CODEOWNERS

aangelisc committed:
Fix gen-cue issue

aangelisc committed:
Fix backend test

aangelisc committed:
Fix e2e test

aangelisc committed:
Update code coverage

aangelisc committed:
Remove references to old Azure Monitor path

andresmgot commented on public/app/plugins/datasource/azuremonitor/types/templateVariables.ts:
```diff
@@ -0,0 +1,14 @@
+export {
+  GrafanaTemplateVariableQueryType,
+  UnknownQuery,
+  AppInsightsMetricNameQuery,
+  AppInsightsGroupByQuery,
+  SubscriptionsQuery,
+  ResourceGroupsQuery,
+  ResourceNamesQuery,
+  MetricNamespaceQuery,
+  MetricDefinitionsQuery,
+  MetricNamesQuery,
+  WorkspacesQuery,
+  GrafanaTemplateVariableQuery,
+} from '../dataquery.gen';
```
being strictly these definitions are frontend-only so in theory we don't need to define them in CUE schemas. Is there any oficial position for this kind of definitions?

aangelisc replied:
I'm not sure there is at the moment, I've defined these temporarily but with the aim of removing as many of them as possible 😊 Phase 2 of the Azure schema work is to massively clean up the types 😜

andresmgot commented on public/app/plugins/datasource/azuremonitor/dataquery.cue:
```diff
@@ -0,0 +1,214 @@
+									subscription:    string
+									resourceGroup:   string
+									resourceName:    string
+									metricNamespace: string
+						}                 @cuetsy(kind="interface")
```
it'd be nice to have some pretiffier here, the indentation and spaces can be a bit arbitrary

Clarity-89 replied:
Apparently there's `make fix-cue` command you can run.

aangelisc replied:
This doesn't seem to do anything for me. Afaict this is the formatted version :/

andresmgot reviewed (approved):
I don't see anything wrong here. Tested and worked for me so :+1:  good job here!

Clarity-89 commented on public/app/plugins/datasource/azuremonitor/types/query.ts:
```diff
@@ -0,0 +1,19 @@
+ * It can represent new queries that are still being edited, so all properties are optional
+ */
+export interface AzureMonitorQuery extends AzureMonitorQueryBase {
+  queryType?: AzureQueryType;
```
Can we also document that it's a workaround until the type extensions in cue are fixed? A TODO comment would be great, so we remember to fix it later.

aangelisc replied:
Yes I'll make sure to do this!

Clarity-89 commented on public/app/plugins/datasource/azuremonitor/dataquery.cue:
```diff
@@ -0,0 +1,214 @@
+							resource?: string
+						} @cuetsy(kind="interface")
+
+						#ResultFormat: "table" | "time_series" @cuetsy(kind="enum")
```
```suggestion
						#ResultFormat: "table" | "time_series" @cuetsy(kind="enum",memberNames="Table|TimeSeries")
```
Looks like we need to provide the member names as the ones with underscores (also dashes) aren't properly formatted.

Clarity-89 commented on public/app/plugins/datasource/azuremonitor/dataquery.gen.ts:
```diff
@@ -0,0 +1,303 @@
+
+export type GrafanaTemplateVariableQuery = (AppInsightsMetricNameQuery | AppInsightsGroupByQuery | SubscriptionsQuery | ResourceGroupsQuery | ResourceNamesQuery | MetricNamespaceQuery | MetricDefinitionsQuery | MetricNamesQuery | WorkspacesQuery | UnknownQuery);
+
+export interface AzureMonitor {}
```
So the actual data source query generated from schema is not used?

aangelisc replied:
I can use the type, however for the sake of readable code the name `AzureMonitor` doesn't make sense. This name should be `AzureMonitorQuery` but I can't tell how to rename the default.

Clarity-89 replied:
@sdboyer shouldn't data query names indeed end in "Query"?

Clarity-89 reviewed (approved):
Great work! 🎉  Left a few minor comments but nothing blocking.

aangelisc commented on public/app/plugins/datasource/azuremonitor/components/LogsQueryEditor/FormatAsField.tsx:
```diff
@@ -4,21 +4,21 @@ import { SelectableValue } from '@grafana/data';
-  { label: 'Time series', value: 'time_series' },
-  { label: 'Table', value: 'table' },
+const FORMAT_OPTIONS: Array<SelectableValue<ResultFormat>> = [
+  { label: 'Time series', value: ResultFormat.Time_series },
```
Yep I'll have a look at this

aangelisc committed:
Review

aangelisc committed:
Merge branch 'main' into schema/azure-monitor

# Conflicts:
#	.drone.yml
#	pkg/kindsys/report.json

aangelisc committed:
Regen files

aangelisc commented:
Merging this and completing #62696. For a first phase schema implementation I'd say this is pretty much there 😊
